dizoo/evogym/envs/evogym_env.py

```diff
@@ -0,0 +1,174 @@
+from typing import Any, Union, List, Optional
+import copy
+import numpy as np
+from easydict import EasyDict
+import gym
+import evogym.envs
+from evogym import WorldObject, sample_robot
+from .viewer import DingEvoViewer
+from evogym.sim import EvoSim
+import os
+from ding.envs import BaseEnv, BaseEnvTimestep, FinalEvalRewardEnv
+from ding.envs.common.common_function import affine_transform
+from ding.torch_utils import to_ndarray, to_list
+from ding.utils import ENV_REGISTRY
+
+
+@ENV_REGISTRY.register('evogym')
+class EvoGymEnv(BaseEnv):
+
+    @classmethod
+    def default_config(cls: type) -> EasyDict:
+        cfg = EasyDict(copy.deepcopy(cls.config))
+        cfg.cfg_type = cls.__name__ + 'Dict'
+        return cfg
+
+    config = dict(
+        env_id='Walker-v0',
+        robot='speed_bot',  # refer to 'world data' for more robots configurations
+        robot_h=5,  # only used for random robots
+        robot_w=5,  # only used for random robots
+        robot_pd=None,  # only used for random robots, probability distributions of randomly generated components)
+        robot_dir=""  # only used for defined robots, path to the robot config, env/world_data/my_bot.json
+    )
+
+    def __init__(self, cfg: dict) -> None:
+        self._cfg = cfg
+        self._init_flag = False
+        self._replay_path = None
+        if 'robot_dir' not in self._cfg.keys():
+            self._cfg = '../'
+
+    def reset(self) -> np.ndarray:
+        if not self._init_flag:
+            self._env = self._make_env()
+            self._env.observation_space.dtype = np.float32  # To unify the format of envs in DI-engine
+            self._observation_space = self._env.observation_space
+            self.num_actuators = self._env.get_actuator_indices('robot').size
+            # by default actions space is double (float64), create a new space with type of type float (float32)
+            self._action_space = gym.spaces.Box(
```

Comment:
It would be better to init self._action_space/_observation_space/... in __init__.
So that no need to assign value during reset.
An env object may reset many times, you know......

Reply:
the problem is that in init, since the environment itself has not yet been instantiated, we can't know the action_space, observation_space, ...